### ISSUE TITLE
syncdir, or something like it

### ISSUE THREAD

**Issue by schmonz** [notqmail/notqmail#94]
According to [DJB's reliability FAQ](https://cr.yp.to/qmail/faq/reliability.html):

> qmail's queue, except for bounce message contents, is crashproof on the BSD FFS and most of its variants.
> 
> Do not use async or softupdates filesystems. If you do, and if your system crashes at the wrong moment, you will lose mail. Under Linux, make sure that all mail-handling filesystems are mounted sync. The same comments apply to many other popular MTAs. (However, some MTAs are unreliable no matter what filesystem you use.)
> 
> It is safe to put qmail's queue on a noatime filesystem.

According to [qmail.org](http://qmailorg.schmonz.com/top.html#addons):

> Bruce Guenter's [syncdir](http://untroubled.org/syncdir/) gives qmail bsd fsync semantics on a Linux filesystem.

syncdir is a library providing replacement implementations of open, link, unlink, and rename that wrap the real system calls and also invoke fsync.

In pkgsrc, which targets a couple dozen Unix platforms, mail/qmail simply takes an unconditional dependency on devel/syncdir. (For portability, I've libtoolized syncdir and provided an alternate implementation using dlsym() for platforms like macOS where syscall() is deprecated or gone.) Some platforms might not need this, but I chose to err on the side of safety.

For notqmail, since people will install on a wide variety of systems, I also think we ought to default to safety. Some options for how to do that:

1. Make syncdir an optional prerequisite for the build, defaulting to failing the build if it's not present
2. Vendor a copy of syncdir into our tree, enabled by default
3. Provide our own fsync-calling wrapper functions that don't shadow the syscall names, and adjust all callers to go through our wrappers

It's probably reasonable to provide a compile-time option for folks who are sure they don't want all the extra fsyncs.

Caveat: syncdir is GPL. This is the first bit of GPL code we might want to include, but it won't be the last (I'm aware of at least [qmail-spp, slated for 1.9](https://github.com/notqmail/notqmail/wiki/Roadmap#19)). I don't know how GPL interacts with public domain. If need be, we can ask Bruce Guenter whether he'll grant us different conditions for use of his code.

**Comment by josuah** [comment]
From far away, this looks like a high cost high benefit change.  Secure... *And safe* (;P) by default is objectively good. Adding different licence and introducing more portability challenges (getting close to syscall interface) require more attention.

If there is an easy way to integrate it, maybe taking the few portable percents that provide what we need, then both goals may be acheived at once without too much compromise.

Let's see what is inside that syncdir...

One-liner to test quickly:

```
tmp=$(mktemp -d); cd "$tmp"; curl https://untroubled.org/syncdir/syncdir-1.0.tar.gz | gzip -cd - | tar -xf -; cd *; cc syncdir.c; cd; rm -rf "$tmp"
```

Alpine Linux no problem, I guess GNU systems are ok too.

OpenBSD for lacks the `<syscall.h>` header:

```
syncdir.c:28:10: fatal error: 'syscall.h' file not found
#include <syscall.h>
         ^~~~~~~~~~~
1 error generated.
```

**Comment by josuah** [comment]
Was the point to avoid too much code change by preserving  the syscall names?

To avoid the need of the <syscall.h> wrapper (other people may have other wrapper, and `<syscall.h>` decreases portability), would you think it would be possible to have the same feature by having wrappers of different names ?

Like `sync_open()`, `sync_rename()`, `sync_link()`, ..., that could call fsync + the system-provided syscall wrapper.

Other solution: call fsync every time in-place, but I suppose wrapper would provide a nicer interface...

**Comment by DerDakon** [comment]
A bit of dump from an IRC discussion, in favor of doing it on our own code if we need to:

* I would _seriously_ avoid any fiddling with syscalls. This will only go wrong. I often enough see it like "hey, let's do special clone()!" and then things explode because sparc puts things differently in registers
* the fact is that things are actually already wrapped, through the things in open.h, we could easily hook at least every open code there
* then there is rename(), which I had to forward-declare because of the stdio.h issue with puts. It's only needed in 2 files. Adding a proper rename_sync(), putting that in a header and then "#include <stdio.h>" in rename_sync.c would look much cleaner to me
* so it boils down to link() and unlink(), and together there are 19 uses, one of them in try*.c, and the otheres in relatively few files. Looks like wrapping them would not make things severely worse
*  we have the bonus that we still can avoid any call to the sync variants at places where we know that it's irrelevant

**Comment by schmonz** [comment]
@josuah, yes to all your analysis. The “minimize code change” goal, in particular, dictated the design of syncdir. For pkgsrc, which supports many OS platforms, I [patched in a fallback implementation with `dlsym()`](https://schmonz.com/qmail/syncdirdlsym/).

Before the Qualys vulnerabilities, “minimize code change” was a goal we strongly shared. But as of 1.08, and continuing toward 1.09, if there’s any tradeoff between defending qmail’s reputation-for-reliability-and-trustworthiness or worrying about how much it costs for qmail admins to rebase their patchset, we’re prioritizing the former.

So I agree with both of you: we don’t need to use syncdir’s implementation to get `fsync()`s where we need them. We can and should accomplish the same effect via more direct, more obvious, and perhaps more invasive-to-custom-patches changes.

**Comment by DerDakon** [comment]
I looked at the places where we have ```rename()``` today. There is qmail-newmrh, qmail-newu and maildir2mbox. All of them use it to exchange a temporary file with the real file. I don't think we need an explicit sync here, either you get the old or the new file, in both cases things are fine. And then there is qmail-pop3d, where it is used to move files from new/ to cur/ to mark them read. But qmail-pop3d tolerates that the rename fails and simply ignores the returrn code as it can't do anything against it. So 4 out of 4 places that use rename() IMHO don't need any sync afterwards.

**Comment by mbhangui** [comment]
Replying to old discussion, but thought I would mention something about compiling syncdir.c on RaspberryPI (with the latest headers) and on FreeBSD.

1.  on the latest linux kernel the syscalls open, link, unlink, rename have been repaced by openat, linkat, unlinkat, renameat.
2. on FreeBSD, one has to include sys/syscall.h

I made tiny changes to syncdir.c to make it work on RPI.

```
#include <sys/types.h>
#include <sys/stat.h>
#define open XXX_open
#include <fcntl.h>
#undef open
#include <unistd.h>
#include <string.h>
#if defined(linux)
#include <syscall.h>
#elif defined(__FreeBSD__)
#include <sys/syscall.h>
#else
#error "syscall.h presence unknown. edit syncdir.c"
#endif
#include <errno.h>
#include <unistd.h>
#include "env.h"

#if defined(SYS_openat) && defined(AT_FDCWD)
#define SYS_OPEN(FILE,FLAG,MODE) syscall(SYS_openat,AT_FDCWD,FILE,FLAG,MODE)
#else
#define SYS_OPEN(FILE,FLAG,MODE) syscall(SYS_open,FILE,FLAG,MODE)
#endif
#define SYS_CLOSE(FD) syscall(SYS_close, FD)
#if defined(SYS_linkat) && defined(AT_FDCWD)
#define SYS_LINK(OLD,NEW) syscall(SYS_linkat,AT_FDCWD,OLD,AT_FDCWD,NEW,0)
#else
#define SYS_LINK(OLD,NEW) syscall(SYS_link,OLD,NEW)
#endif
#if defined(SYS_unlinkat) && defined(AT_FDCWD)
#define SYS_UNLINK(PATH) syscall(SYS_unlinkat,AT_FDCWD,PATH,0)
#else
#define SYS_UNLINK(PATH) syscall(SYS_unlink,PATH)
#endif
#if defined(SYS_renameat) && defined(AT_FDCWD)
#define SYS_RENAME(OLD,NEW) syscall(SYS_renameat,AT_FDCWD,OLD,AT_FDCWD,NEW,0)
#else
#define SYS_RENAME(OLD,NEW) syscall(SYS_rename,OLD,NEW)
#endif
#define SYS_FSYNC(FD) syscall(SYS_fsync, FD)
```

**Comment by schmonz** [comment]
Summarizing and reminding ourselves: the path forward here is emphatically _not_ syncdir. That would be a relatively complicated fix. Since we're the actively maintained upstream, we instead will directly add `fsync()` calls where needed.

**Comment by schmonz** [comment]
Before syncdir became Unlicensed, I worked up a proof-of-concept branch integrating its logic directly. Now that [syncdir is Unlicensed](https://untroubled.org/syncdir/), I'll rework that branch and perhaps propose a PR.

**Comment by schmonz** [comment]
Related, perhaps to be solved separately:

> As it turns out, macOS cheats. On Linux, fsync() will both flush writes to the drive, and ask it to flush its write cache to stable storage.
> 
> But on macOS, fsync() only flushes writes to the drive. Instead, they provide an F_FULLSYNC operation to do what fsync() does on Linux.

(from https://twitter.com/marcan42/status/1494213862970707969)
